Schedule anytime delay in the console

pgalic96 opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Closes #1840 

#### Changes
<!-- What are the changes made in this pull request? -->

- Add new `ScheduleDelayInput` component that is composed of two fields
- Add `schedule_anytime_delay` field when making api request

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

- Since `schedule_anytime_delay` is returned as a string, I added some logic for checking substrings for unit and value. This was a bit tricky when it came to validating and especially displaying warnings when the value is lower than `130ms` . Initially I wanted to make the `ScheduleDelayInput` more generic, to be able to add it to other places that possibly need a duration input, but didn't find a better way of validation outside of this component, so the component is now quite specific when checking allowed values and such things. If you have an idea how to do the validation of this string value outside of the component let me know

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [x] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [x] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

kschiffer commented on pkg/webui/console/containers/schedule-delay-input/schedule-delay-input.styl:
```diff
@@ -0,0 +1,43 @@
+      margin-bottom: 0
+
+    .number
+      padding-right 2rem
```
```suggestion
      padding-right $cs.s
```
Always use scale variables.

kschiffer commented on pkg/webui/console/containers/schedule-delay-input/index.js:
```diff
@@ -0,0 +1,63 @@
+import Field from '../../../components/form/field'
+import DurationInput from './schedule-delay-input'
+
+const ScheduleDelayField = function({
```
Unless I'm not missing something here, this whole construct is unnecessary.
You could simply use the `<Field />` component directly and pass the props. Wrapping only makes sense if you need something from the store to be present in the field.

kschiffer commented on pkg/webui/console/components/gateway-data-form/index.js:
```diff
@@ -160,6 +164,12 @@ class GatewayDataForm extends React.Component {
+          name="schedule_anytime_delay"
+          description={m.scheduleAnyTimeDescription}
+          required
+        />
```
```suggestion
        <Field
          component={ScheduleDelayInput}
          title={m.scheduleAnyTimeDelay}
          name="schedule_anytime_delay"
          description={m.scheduleAnyTimeDescription}
          required
        />
```

kschiffer commented on pkg/webui/console/containers/schedule-delay-input/schedule-delay-input.js:
```diff
@@ -0,0 +1,158 @@
+  { label: m.hours, value: 'h' },
+]
+
+class ScheduleDelayInput extends React.PureComponent {
```
This is not a container component, you're not connecting to the store or fetch anything.
Move this to `/components`

kschiffer commented on pkg/webui/console/views/gateway/index.js:
```diff
@@ -91,6 +91,7 @@ import PropTypes from '../../../lib/prop-types'
       'auto_update',
       'schedule_downlink_late',
       'update_channel',
+      'schedule_anytime_delay',
```
Does it make sense to show this value also in the gateway overview? @htdvisser

htdvisser replied:
I don't have an opinion on this.

kschiffer replied:
Let's say no then.

pgalic96 commented:
Updated the component to be generic, the only thing I am not sure is about setting the state of `shouldDisplayWarning` , and whether that could be separated from `encode` and `decode` functions cc @kschiffer let me know

kschiffer commented on pkg/webui/console/components/duration-input/index.js:
```diff
@@ -0,0 +1,118 @@
+    const { value, decode } = this.props
+
+    const initialDelay = decode(value)
+    this.state = {
```
You don't need to keep state of the value. The value prop passed to the component is sufficient and you can derive all the values you need from it. I propose to use the `withComputedProps` HOC, like this:
```js
@withComputedProps(props => ({
  ...props,
  value: props.decode(props.value),
}))
class DurationInput extends React.PureComponent {
   …
```
This way, the value prop will always be in the form that you need and you can encode it again before passing it to the `onChange` prop.

Since the component is quite simple and only uses the value prop in the render function, you could alternatively just pipe the value through the decoder function there.

kschiffer commented on pkg/webui/console/components/duration-input/index.js:
```diff
@@ -0,0 +1,118 @@
+import PropTypes from '../../../lib/prop-types'
+import style from './duration-input.styl'
+
+class DurationInput extends React.PureComponent {
```
```suggestion
class UnitInput extends React.PureComponent {
```
It doesn't have to be used for durations only.

kschiffer commented on pkg/webui/console/components/gateway-data-form/index.js:
```diff
@@ -44,6 +48,14 @@ const m = defineMessages({
+  minutes: 'minutes',
+  hours: 'hours',
+  delayWarning:
+    'Value of delay too low. The lower bound (130ms) will be used by the Gateway Server.',
```
```suggestion
    'Delay too short. The lower bound (130ms) will be used by the Gateway Server.',
```

kschiffer commented on pkg/webui/console/components/gateway-data-form/index.js:
```diff
@@ -69,19 +81,74 @@ const validationSchema = Yup.object().shape({
+    const duration = value.split(unitRegexp)[0]
+    const unit = value.split(duration)[1]
+
+    this.setState({ shouldDisplayWarning: this.isNotValidDuration(duration, unit) })
```
While it can be done like this too, I think the cleaner approach is to hook into the `onChange` event of the Field component and to check the value there.

```js
@bind
onScheduleAnytimeDelayChange(value) {
  this.setState({ shouldDisplayWarning: this.isNotValidDuration(value) }) // also add the decoding to the check function then
}
```

kschiffer commented on pkg/webui/console/components/gateway-data-form/index.js:
```diff
@@ -69,19 +81,74 @@ const validationSchema = Yup.object().shape({
+  schedule_anytime_delay: Yup.string().required(sharedMessages.validateRequired),
 })
 
 @bind
```
Can you refactor the class bind to method bindings while at it?

kschiffer commented on pkg/webui/console/components/gateway-data-form/index.js:
```diff
@@ -160,6 +227,22 @@ class GatewayDataForm extends React.Component {
+          name="schedule_anytime_delay"
+          component={DurationInput}
+          description={m.scheduleAnyTimeDescription}
+          units={[
```
```suggestion
          onChange={this.onScheduleAnytimeDelayChange}
          units={[
```

kschiffer commented on pkg/webui/console/components/gateway-data-form/index.js:
```diff
@@ -69,19 +81,74 @@ const validationSchema = Yup.object().shape({
+  }
+
+  isNotValidDuration(duration, unit) {
+    const durationLowerBound = 130 // set lower bound duration in ms
```
Is this value going to stay that way, or should it be fetched somehow from the backend? If it stays like this, I'd suggest to add it to our `pkg/webui/console/constants`.

pgalic96 replied:
it cannot be fetched from the backend I believe, because it's hidden somewhere deep in GS. I will add it to constants

pgalic96 commented:
I implemented the suggested changes. I had to change the validation of the field, and handling of empty values because the regex splitting worked only if both delay and unit are present. Before it worked because we kept a state of variables, but now if unit was undefined, there was no way of recovering the previous value of unit since its calculated from props only. Also I added some regular expressions to handle these special cases.

kschiffer commented on pkg/webui/console/lib/regexp.js:
```diff
@@ -24,3 +24,6 @@ export const mqttPassword = /^(?![\s\S])|.{2,100}/
 export const latitude = /^[-+]?([1-8]?\d(\.\d+)?|90(\.0+)?)$/
 export const longitude = /^\s*[-+]?(180(\.0+)?|((1[0-7]\d)|([1-9]?\d))(\.\d+)?)$/
 export const int32 = /^-?\d+$/
+export const unit = new RegExp('[a-z]{1,}')
```
```suggestion
export const unit = new RegExp('[a-zA-Z]{1,}')
```

kschiffer commented on pkg/webui/console/constants/delays.js:
```diff
@@ -0,0 +1,17 @@
+// limitations under the License.
+
+export default Object.freeze({
+  GATEWAY_SCHEDULE: 130, // miliseconds
```
Also, please add the default value here.

kschiffer commented on pkg/webui/console/components/unit-input/index.js:
```diff
@@ -0,0 +1,114 @@
+            step="any"
+            name={name}
+            onBlur={this.handleBlur}
+            value={duration}
```
```suggestion
            value={duration}
            required={required}
```
Pass on the required prop to the input component.

pgalic96 commented on pkg/webui/console/components/gateway-data-form/index.js:
```diff
@@ -44,6 +51,13 @@ const m = defineMessages({
   updateChannelDescription: 'Channel for gateway automatic updates',
   enforceDutyCycleDescription:
     'Recommended for all gateways in order to respect spectrum regulations',
+  scheduleAnyTimeDelay: 'Schedule Any Time Delay',
```
not sure about this, @neoaggelos ?

pgalic96 commented:
Implemented all your feedback and tested, will rebase once its approved
